### PR TITLE
cql3: castas_fcts: do not rely on boost casting large multiprecision integers to floats behavior

### PR DESCRIPTION
**Pull request by avikivity**
In [1] a bug casting large multiprecision integers to floats is documented (note that it
received two fixes, the most recent and relevant is [2]). Even with the fix, boost now
returns NaN instead of ±∞ as it did before [3].

Since we cannot rely on boost, detect the conditions that trigger the bug and return
the expected result.

The unit test is extended to cover large negative numbers.

Boost version behavior:
 - 1.78 - returns ±∞
 - 1.79 - terminates
 - 1.79 + fix - returns NaN

Fixes https://github.com/scylladb/scylladb/issues/18508

[1] https://github.com/boostorg/multiprecision/issues/553
[2] https://github.com/boostorg/multiprecision/commit/ea786494db78efdf178cbe36810f3326156e3347
[3] https://github.com/boostorg/math/issues/1132

- [x] no backport - bug only affects toolchains based on Fedora 40; no branches affected

### CODE REVIEW

**Comment by nyh** [comment]
> In [1] a bug casting large multiprecision integers to floats is documented. Until the fix propagates, work around it by detecting the conditions that trigger the bug and returning the expected result.
> 
> [1] [boostorg/multiprecision#553](https://github.com/boostorg/multiprecision/issues/553)

@tchaikov now I'm confused -  according to [1], this was a known bug, and was solved a year ago. So why did you need to solve it again? Did they just fail to check that their fix actually fixed the bug - and it didn't? Or was the bug solved and then reappeared? (if that the case then perhaps referring to [1] now is misleading?).

**Review by nyh — APPROVED** [review]
The fix looks good, but I would just change the comment referencing the issue https://github.com/boostorg/multiprecision/issues/553 - this issue was already closed and fixed a year (!) ago, so why is it affecting us now? How will we know in the future when we can remove this workaround, if the closure of the references issue doesn't mean it was fixed?

**Comment by tchaikov** [comment]
> > In [1] a bug casting large multiprecision integers to floats is documented. Until the fix propagates, work around it by detecting the conditions that trigger the bug and returning the expected result.
> > [1] [boostorg/multiprecision#553](https://github.com/boostorg/multiprecision/issues/553)
> 
> @tchaikov now I'm confused - according to [1], this was a known bug, and was solved a year ago. So why did you need to solve it again? 

i'd say the fix was not good enough. or there was a regression after the fix landed.

> Did they just fail to check that their fix actually fixed the bug - and it didn't? Or was the bug solved and then reappeared? (if that the case then perhaps referring to [1] now is misleading?).

it's not misleading. if i were the maintainer, i would reopen it and then close it.

**Comment by tchaikov** [comment]
> The fix looks good, but I would just change the comment referencing the issue [boostorg/multiprecision#553](https://github.com/boostorg/multiprecision/issues/553) - this issue was already closed and fixed a year (!) ago, so why is it affecting us now? 

because we want to use a newer Boost packaged by fedora 40. and this version of Boost happens to include this regression.

> How will we know in the future when we can remove this workaround, if the closure of the references issue doesn't mean it was fixed?

i am tracking the PR at https://src.fedoraproject.org/rpms/boost/pull-request/24 . once it's merged. i will replace this workaround with a slightly better fix. see also https://github.com/scylladb/scylladb/issues/18508#issuecomment-2094541964

**Review by tchaikov — APPROVED** [review]
lgtm.

**Comment by avikivity** [comment]
v2: expanded commit message

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
> i am tracking the PR at https://src.fedoraproject.org/rpms/boost/pull-request/24 . once it's merged. i will replace this workaround with a slightly better fix. see also [#18508 (comment)](https://github.com/scylladb/scylladb/issues/18508#issuecomment-2094541964)

If the fixed Boost now returns NaN instead of +Inf/-Inf, it seems to me you'll still need some sort of workaround on our end if we want the cast behavior not to change :-(

**Comment by tchaikov** [comment]
> > i am tracking the PR at https://src.fedoraproject.org/rpms/boost/pull-request/24 . once it's merged. i will replace this workaround with a slightly better fix. see also [#18508 (comment)](https://github.com/scylladb/scylladb/issues/18508#issuecomment-2094541964)
> 
> If the fixed Boost now returns NaN instead of +Inf/-Inf, it seems to me you'll still need some sort of workaround on our end if we want the cast behavior not to change :-(

yeah, i knew, that's the 3rd step.

**Comment by nyh** [comment]
> > If the fixed Boost now returns NaN instead of +Inf/-Inf, it seems to me you'll still need some sort of workaround on our end if we want the cast behavior not to change :-(
> 
> yeah, i knew, that's the 3rd step.

Oh, right, I missed that.

Please add to test/cql-ptest/test_cast_data.py also a test for -inf, because I assume very negative numbers would be cast (by older Scylla and by Cassandra) to -Inf, not to Inf, and I didn't test that - I only tested the +inf case (which was a lapse in judgement on my part).

**Comment by avikivity** [comment]
Ah, so it's not just a workaround for a bug, it's required for changed behavior. Now I understand your 3-step plan better.

The changelog needs adjustment to reflect this.

Was the change to NaN intentional?

**Comment by tchaikov** [comment]
> Was the change to NaN intentional?

i was not the one who made this call. neither was boost.multiprecision. i think it was boost.math's author made this decision. but i have not read the implementation in great details. so i am not 100% sure.

what i did was just to ensure the `noexcept` function is noexcept. that is the right behavior, i believe.

**Comment by nyh** [comment]
> > Was the change to NaN intentional?
> 
> i was not the one who made this call. neither was boost.multiprecision. i think it was boost.math's author made this decision. but i have not read the implementation in great details. so i am not 100% sure.

I'm not happy about the development process the Boost guys used here. They had a known bug, https://github.com/boostorg/multiprecision/issues/553, that was supposedly fixed without adding a regression test so somehow it regressed in less than a year (if it was ever fixed?). And they now they quickly merged a patch that converted the very serious bug (crash) to a less-serious-but-still-a-bug (NaN) and for lack of testing didn't notice that fact. When I warned of this NaN change during the review (https://github.com/boostorg/multiprecision/pull/618), my worry was ignored by the developer. Maybe my worry is wrong, but they didn't explain why (and I don't think it's wrong, wrapping a huge number to NaN doesn't make sense). Not confidence-inspiring.

By the way, if you're wondering why I thought a huge integer should wrap to Inf, not to NaN, well, this is not only the old behavior of this code a year ago (so it shouldn't change because it breaks backward compatibility!), it's also the tradition of what happens when overflowing floating point. For example:
```
$ python
>>> print(1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0)
inf
```

**Comment by avikivity** [comment]
v2:
 - rebased
 - expanded changelog 
 - expanded unit test for -inf

**Review by nyh — APPROVED** [review]

**Pull request closed without merging**
